libs/langchain-azure-cosmosdb/src/azure_cosmosdb_nosql.ts

```diff
@@ -0,0 +1,522 @@
+            { name: "@embeddingKey", value: this.embeddingKey },
+            { name: "@vector", value: queryVector },
+          ],
+        },
+        { maxItemCount: k }
+      )
+      .fetchAll();
+
+    const docsAndScores = items.map(
+      (item) =>
+        [
+          new Document({
+            id: item.id,
+            pageContent: item.text,
+            metadata: {
+              ...(item.metadata ?? {}),
+              ...(filter?.includeEmbeddings
+                ? { [this.embeddingKey]: item.vector }
+                : {}),
+            },
+          }),
+          item.similarityScore,
+        ] as [Document, number]
+    );
+
+    return docsAndScores;
+  }
+
+  /**
+   * Return documents selected using the maximal marginal relevance.
+   * Maximal marginal relevance optimizes for similarity to the query AND
+   * diversity among selected documents.
+   * @param query Text to look up documents similar to.
+   * @param options.k Number of documents to return.
+   * @param options.fetchK=20 Number of documents to fetch before passing to
+   *     the MMR algorithm.
+   * @param options.lambda=0.5 Number between 0 and 1 that determines the
+   *     degree of diversity among the results, where 0 corresponds to maximum
+   *     diversity and 1 to minimum diversity.
+   * @returns List of documents selected by maximal marginal relevance.
+   */
+  async maxMarginalRelevanceSearch(
+    query: string,
+    options: MaxMarginalRelevanceSearchOptions<this["FilterType"]>
+  ): Promise<Document[]> {
+    const { k, fetchK = 20, lambda = 0.5 } = options;
+    const includeEmbeddingsFlag = options.filter?.includeEmbeddings || false;
+
+    const queryEmbedding = await this.embeddings.embedQuery(query);
+    const docs = await this.similaritySearchVectorWithScore(
+      queryEmbedding,
+      fetchK,
+      {
+        ...options.filter,
+        includeEmbeddings: true,
+      }
+    );
+    const embeddingList = docs.map((doc) => doc[0].metadata[this.embeddingKey]);
+
+    // Re-rank the results using MMR
+    const mmrIndexes = maximalMarginalRelevance(
+      queryEmbedding,
+      embeddingList,
+      lambda,
+      k
+    );
+
+    return mmrIndexes.map((index) => {
+      const doc = docs[index][0];
+
+      // Remove embeddings if they were not requested originally
+      if (!includeEmbeddingsFlag) {
+        delete doc.metadata[this.embeddingKey];
+      }
+      return doc;
+    });
+  }
+
+  /**
+   * Initializes the AzureCosmosDBNoSQLVectorStore by connecting to the database.
+   * @param client The CosmosClient to use for connecting to the database.
+   * @param databaseName The name of the database to use.
+   * @param containerName The name of the collection to use.
+   * @param initOptions Initialization options for the database and container.
+   * @returns A promise that resolves when the AzureCosmosDBNoSQLVectorStore has been initialized.
+   */
+  private async init(
```

Comment:
Might be good to expose this as `initialize`?

Reply:
Done. I've added a public `initialize` method, and defer its call to the first time it's needed, so it can either be explicitly called after constructor, or it will be called automatically when any of the other methods needs it.